Enhancement of patient details page

Mahendar0701 opened the pull request:
## Proposed Changes

- Fixes #8808 
- Implemented some parts of proposed design of patient details page
- Created tabs for Patient profile, Consultation history, Test sample history, Shifting history, Visit history as suggested
- Shifted add consultation blocks to right side and made it responsive

Proposed:
![image](https://github.com/user-attachments/assets/e4b4a918-3676-4062-ae2b-b376dddb59ea)

Modified page:
![image](https://github.com/user-attachments/assets/e01b0137-2aa6-4b74-bdcb-cbfd1f3f993d)

Need further suggestion on implementing ABHA card and Hospital ID Card


@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [ ] Request for Peer Reviews
- [ ] Completion of QA


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

- **New Features**
  - Enhanced routing capabilities for patient-related views, including new pages for demographics, encounters, health profiles, immunization records, sample test history, and shifting history.
  - Introduction of a notes management component for patient notes.

- **Improvements**
  - Updated layout and styling for various components, improving user experience and responsiveness.
  - Enhanced localization support with new keys for multiple languages, including English, Hindi, Kannada, Malayalam, and Marathi.
  - Improved navigation within the patient registration form based on URL parameters.
  - Adjusted visual elements, including breadcrumb styling and patient information displays.
  - Enhanced the `PatientHome` component with a new tabbed interface for better organization of patient details.
  - Streamlined the `ShiftingBlock` component for optional transfer handling.
  - Improved clarity and accuracy of information displayed in the `PatientInfoCard` and `InsuranceDetails` components.
  - Updated the `Chip` component's styling for better visual consistency.
  - Improved the user management interface with better validation and error handling.

- **Bug Fixes**
  - Improved error handling and validation in user management and patient registration components.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Mahendar0701 committed:
adjusted ui of patient details

Mahendar0701 committed:
shifted blocks to right

Mahendar0701 committed:
small changes

Mahendar0701 committed:
solved merge conflicts

Mahendar0701 committed:
solved merge conflicts

samholics commented:
<img width="262" alt="image" src="https://github.com/user-attachments/assets/5b768e25-30e6-41f8-9869-fa2c3fd6c6b7">

- [ ]  covid vaccination details may come under covid details heading
or 
- [ ] Rename No.of vaccine doses  to no.of Covid Vaccine doses
since the covid out of context now, this will confuse people with what vaccine doses are mentioned here.

khavinshankar commented on src/Components/Patient/InsuranceDetailsCard.tsx:
```diff
@@ -12,12 +12,12 @@ export const InsuranceDetialsCard = (props: InsuranceDetails) => {
         </div>
         {data ? (
-          <div className="grid grid-cols-1 gap-2 sm:grid-cols-2 md:grid-cols-2">
+          <div className="border-gray-150 grid grid-cols-1 gap-2 rounded-md border bg-white p-5 shadow sm:grid-cols-2 md:grid-cols-2">
```
instead of directly using gray, use secondary, `border-secondary-100`

Mahendar0701 replied:
Yes i have updated

bodhish commented on src/Components/Common/Avatar.tsx:
```diff
@@ -34,6 +34,9 @@ const toColor = (name: string): [string, string] => {
 };
 
 const initials = (name: string): string => {
+  if (!name) {
+    return ""; // Return empty string or a default value if the name is undefined or null
+  }
```
When would this happen 🤔

Mahendar0701 replied:
At a certain point, I encountered an issue, so i added it. But now its fine. I will remove it.

khavinshankar commented on src/Components/Common/Avatar.tsx:
```diff
@@ -34,6 +34,9 @@ const toColor = (name: string): [string, string] => {
 const initials = (name: string): string => {
+  if (!name) {
+    return ""; // Return empty string or a default value if the name is undefined or null
+  }
```
Is this even required, for this function we expect our inputs to be a string, incase if there is a possibility of undefined or null, it must be handled while calling this function, ex: `initials(user?.name ?? "unknown")`.

khavinshankar commented on src/Components/Patient/InsuranceDetailsCard.tsx:
```diff
@@ -12,12 +12,12 @@ export const InsuranceDetialsCard = (props: InsuranceDetails) => {
-          Policy Details
+      <div className="h-full space-y-2 pr-5">
+        <div className="mr-4 w-full pb-2 text-xl font-bold text-secondary-900">
+          Insurance Details
```
i18n

When you work on a file that has text without translations, please feel free to add it.

khavinshankar commented on src/Components/Patient/PatientHome.tsx:
```diff
@@ -376,470 +376,1427 @@ export const PatientHome = (props: any) => {
+                                  <CareIcon icon="l-edit-alt" className="text-lg" />
+                                  Edit
+                                </ButtonV2>
+                              </div> */}
```
remove the comment, unless it might be useful later

khavinshankar commented on src/Components/Patient/SampleDetails.tsx:
```diff
@@ -242,7 +242,7 @@ export const SampleDetails = ({ id }: DetailRoute) => {
       <Card key={flow.id}>
         <div className="grid grid-cols-1 gap-4 md:grid-cols-2">
-          <div>
+          <div className="">
```
```suggestion
          <div>
```

khavinshankar commented on src/Components/Patient/PatientHome.tsx:
```diff
@@ -376,470 +376,1427 @@ export const PatientHome = (props: any) => {
+                  ? "border-b-2 border-green-600 font-semibold text-green-700"
+                  : "rounded-lg hover:bg-gray-100"
+              }`}
+              onClick={() => setActiveTab("consultation-history")}
```
use classNames or cn util to have conditional class names, string interpolation are hard to read

bodhish reviewed (changes requested):
- [ ] Please make the tabs a seperate page; 

Patient profile: .../patient/:id/demographic 
Encounters: .../patient/:id/encounters 
Lab Test: .../patient/:id/lab_tests 
Shifting: .../patient/:id/shifting
Notes: .../patient/:id/notes

- [ ] Remove duplicate alerts 
<img width="1446" alt="image" src="https://github.com/user-attachments/assets/9589e8f3-4ec0-41ba-b324-b1b4fa78835b">
- [ ] Not a fan of the right side, we its now a mix of actions and pages so it should probably be refactored to just action. @vinutv will share the design in few days. 
<img width="603" alt="image" src="https://github.com/user-attachments/assets/0ffafc97-829b-458a-9a22-7d256ddefd48">

- [ ] Broken tags 
<img width="562" alt="image" src="https://github.com/user-attachments/assets/c5b6f79a-9db7-466a-b3c5-ca2ba08243cb">

- [ ] Left side has a lot of white space. @vinutv can you check and share your comments
https://deploy-preview-8856--care-ohc.netlify.app/facility/0c95c7f0-e1d2-4aff-83fa-933cef60d3a8/patient/7e8fc20f-02da-47f7-8c33-f2d8c7e52dd8

Mahendar0701 committed:
added changes

Mahendar0701 committed:
removed unused code

Mahendar0701 committed:
added translations

vinutv commented:
>  * [ ] Not a fan of the right side, we its now a mix of actions and pages so it should probably be refactored to just action. @vinutv will share the design in few days. <img alt="image" width="603" src="https://private-user-images.githubusercontent.com/14979190/378644720-0ffafc97-829b-458a-9a22-7d256ddefd48.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk2OTc1NDgsIm5iZiI6MTcyOTY5NzI0OCwicGF0aCI6Ii8xNDk3OTE5MC8zNzg2NDQ3MjAtMGZmYWZjOTctODI5Yi00NThhLTlhMjItN2QyNTZkZGVmZDQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDIzVDE1MjcyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRhZjFhOTJkNjRiYjE3NWExMTNkMTRmNmVlODJmZjRjNDFhY2UwODRjNzk5MzAyNTFjMzBjZGQ5ZTc4Njg5YTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.bNXQ8eUMpiMnKw3TsUMbSCddD4sk68kvrLxfSda62iQ">

**Here is the updated design:** https://www.figma.com/design/Z93EYKSa1MdBmXndsMBJSQ/Care?node-id=905-901
- Redesigned the patient demography section into a 3-column layout to optimise space on the page.
- Placed navigation links for demography sections in the left column. These links can be hidden on smaller screens for better responsiveness.
- Displayed the demography details in the center column.
- Added quick links in the right column for easy access.
cc @bodhish @Mahendar0701

Mahendar0701 committed:
added changes according to proposed design

Mahendar0701 committed:
Merge branch 'develop' into adjusting-margins-in-patient-details#8808

Mahendar0701 committed:
fixing failed build

Mahendar0701 committed:
fixing build

Mahendar0701 commented:
Hi @bodhish  and @vinutv ,

Here is update on my PR. I have made changes to the PR. Could you please take a look at it?

Additionally, I have a few questions:

What should be included in the "Quick Links" section?
Is the "Encounter History" tab referring to the consultation history?
The patient model doesn't include UHID and ABHA ID. Where can I find them?
The volunteer object doesn't have a contact number. Should I fetch the volunteer's contact number using the volunteer ID?


I have yet to implement the links to the tabs.

bodhish commented:
Overall its shaping up well, good progress @Mahendar0701 

> What should be included in the "Quick Links" section?
Generally it should be actions, keep what you have for now
> Is the "Encounter History" tab referring to the consultation history?
Yeah, call it `Encounters`
> The patient model doesn't include UHID and ABHA ID. Where can I find them?
ABHA ID will only be present if the patient is linked to ABDM. (The UHID design is futuristic, use patient ID as UHID for now) 
> The volunteer object doesn't have a contact number. Should I fetch the volunteer's contact number using the volunteer ID?
Volunteer is a user, you can fetch the details if a volunteer is assigned. (low priority task)
> I have yet to implement the links to the tabs.
Lets get this done, All the tab components will become pages and that will be the major change here. lets also make sure each page will take care of its data loading

Mahendar0701 committed:
Resolved conflicts after applying stash

Mahendar0701 committed:
added links to tabs

Mahendar0701 committed:
merged

Mahendar0701 committed:
untracked ConsultationFeedTab.tsx

bodhish commented:
This will have way more story points for sure; 

Quick code review those are not tabs, those are pages so let's have static paths for each page.

Mahendar0701 committed:
refactored code

Mahendar0701 committed:
Merge branch 'develop' into adjusting-margins-in-patient-details#8808

rithviknishad reviewed (changes requested):
quick code review: let's not have `any` type. define/use proper types.

Mahendar0701 committed:
replaced type any with required one's

Mahendar0701 commented:
Hi @bodhish  can you please review my update. I almost completed the proposed design .

vinutv committed:
Improve styling [WIP]: Add sticky behavior to tab and side navigation links on scroll

Mahendar0701 committed:
implement dynamic sidebar color updates on section scroll

Mahendar0701 committed:
Merge branch 'develop' into adjusting-margins-in-patient-details#8808

Mahendar0701 committed:
added edit buttons to each section with navigation and scroll functionality

vinutv committed:
Improve styling [WIP]: Fix spacing for the 3 column layout

vinutv committed:
Improve styling [WIP]: fix the right column spacing and broken styling on the sticky tab

vinutv committed:
Update colors for the breadcrumbs

vinutv committed:
Improve small-screen styling and label text color consistency

Mahendar0701 committed:
resolved merge conflicts

bodhish commented:
@coderabbitai full review

rithviknishad reviewed (changes requested):
Avoid adding translations for languages other than `en` https://github.com/ohcnetwork/care_fe/#translations


And like already said, `any` type is not allowed. I still see usages of `any` type.

bodhish commented on src/components/Patient/PatientDetailsTab/ImmunisationRecords.tsx:
```diff
@@ -0,0 +1,137 @@
+
+export const ImmunisationRecords = (props: PatientProps) => {
+  const { facilityId, id } = props;
+  const [patientData, setPatientData] = useState<PatientModel>({});
```
I am not sure why we are loading the patient across all file, why can't the parent load this and pass it on as prop?

bodhish commented on src/components/Patient/PatientHome.tsx:
```diff
@@ -307,1108 +224,331 @@ export const PatientHome = (props: any) => {
+            page === tab.route && (
+              <tab.component key={tab.route} facilityId={facilityId} id={id} />
+            ),
+        )}
```
Not a fan, I would rather write a switch statement here; Also make sure that all my possible values are typed over string.

bodhish commented:
@vinutv can you check this out; We should probably trim the tab names also; 
I also have a feeling that the right nav should be preserved across these inside pages, what do you think?

Also @Mahendar0701 the notes should be part of the tab.

vinutv committed:
Adjust z-index for menu component to resolve layering issues

vinutv committed:
Shorten tab names for improved readability on patient page UI

vinutv committed:
Update spacing for the patient profile

vinutv commented:
<img width="288" alt="Screenshot 2024-11-07 at 1 23 41 AM" src="https://github.com/user-attachments/assets/df8fff56-6362-426b-8fca-6969d419d729">
Can we hide actions that are restricted to the user, rather than displaying them as disabled, to prevent confusion?

bodhish commented:
We can hide

vinutv commented:
![Patient-Encounters](https://github.com/user-attachments/assets/3d5777bb-26d3-40a3-a375-0272847fd008)

The right sticky panel should be retitled "Actions" and display only patient-specific actions. All other links should be organised as tabs. There should be a single **primary call-to-action** on to patient card based on the user’s role. For example, for doctors, the CTA should be "Start Consultation", and for front desk staff, it should be "Book Appointment".

What do you think @bodhish ?

vinutv commented:
<img width="768" alt="Screenshot 2024-11-07 at 5 05 19 PM" src="https://github.com/user-attachments/assets/77cad9b4-0cd2-47cd-89bf-8dd916da2a9d">

The alert and information labels should be integrated directly into the patient card. There is no need for an additional "Create Consultation" warning alert. Placing multiple primary CTAs across the page may confuse the user, so a single, clearly defined primary CTA should be used.

Mahendar0701 committed:
added notes tab and removed disabled buttons

vinutv committed:
Update spacing for badges, patient card text style

Mahendar0701 commented:
> ![Patient-Encounters](https://private-user-images.githubusercontent.com/16238306/383938385-3d5777bb-26d3-40a3-a375-0272847fd008.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzEwNzg2OTgsIm5iZiI6MTczMTA3ODM5OCwicGF0aCI6Ii8xNjIzODMwNi8zODM5MzgzODUtM2Q1Nzc3YmItMjZkMy00MGEzLWEzNzUtMDI3Mjg0N2ZkMDA4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTA4VDE1MDYzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjNmFmNmExYWYxZWMzYWQ1MGQwMDJmODM2MTA4MzM4YWExMzM4ZjgzNWI4ZTY1MWJmNmRjODlmMjBiYjZkZWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SMgsHeGqOQl4KV2jPP5L_RPdCIsdRsyRiPSfXzr6yHQ)
> 
> The right sticky panel should be retitled "Actions" and display only patient-specific actions. All other links should be organised as tabs. There should be a single **primary call-to-action** on to patient card based on the user’s role. For example, for doctors, the CTA should be "Start Consultation", and for front desk staff, it should be "Book Appointment".
> 
> What do you think @bodhish ?

Right sticky action panel should displayed on all tabs right?

bodhish commented:
Yep; that will be part of the parent section

Mahendar0701 commented:
@bodhish  @vinutv 
What should be displayed under Shift patient and Request test tab. Currently i am showing Shift history and sample test history . Should i have to show Shift patient form and request sample test form under respective tabs?. Or let it be same?

bodhish commented:
Keep that, let's improve as a next PR.

Mahendar0701 committed:
action panel updated, removed dropdown from patient card and refactored code

Mahendar0701 committed:
added alert and info labels in patient card

Mahendar0701 commented:
@bodhish  @vinutv 

I made the right action panel visible on all pages, added "Notes" as a tab, included a consultation warning message, displayed assigned details in the patient card, and removed the patient actions dropdown. Could you please review these changes and let me know if any improvements are needed?

bodhish commented:
Can you also address CodeRabbit suggestions?

Mahendar0701 committed:
added coderabbit suggestions

Mahendar0701 commented:
Added most of the coderabbit suggestions

Mahendar0701 committed:
resolved merge conflicts

Mahendar0701 committed:
resolved conflicts

nihal467 commented:
@Mahendar0701 is this ready for QA testing ?

Mahendar0701 commented:
> @Mahendar0701 is this ready for QA testing ?

Some parts of design yet to implement. I will update when it is done

shivankacker commented:
@Mahendar0701 please let me know when you are done with all the changes

shivankacker committed:
cleanup

shivankacker committed:
bugfix

shivankacker committed:
fixed merge conflicts

Mahendar0701 commented:
> @Mahendar0701 please let me know when you are done with all the changes

I will update only EncounterHistory.tsx, can you please review all other work

shivankacker committed:
encounters page complete

shivankacker committed:
finished up

shivankacker committed:
commented out ABHA number

shivankacker committed:
tests

shivankacker committed:
tests 2: electric boogaloo

shivankacker committed:
Tests 3 : When will we be free?

shivankacker committed:
Test 4 : This is so bore

shivankacker committed:
Test 5 : Hardly Alive

shivankacker committed:
revert package lock

shivankacker committed:
Merge branch 'develop' of https://github.com/ohcnetwork/care_fe into pr/Mahendar0701/8856

shivankacker committed:
Tests 6: I just wanna fix

shivankacker committed:
Tests 7 : We might go till eleven

shivankacker committed:
Tests 8 : Flaky tests break my rate

shivankacker committed:
Tests 9 : Is this really flaky or does my mind not shine?

Mahendar0701 committed:
added margin between insurance details

bodhish commented:
- [ ] The volunter tab is shown on the left side while its not present in the center 
![image](https://github.com/user-attachments/assets/183b01f5-5a9b-418c-8503-e560f8b47f7b)

- [ ] Show Shifiting details as a list?
![image](https://github.com/user-attachments/assets/e3a3e459-4770-4435-961d-1b91b2eb8ab9)

- [ ] Rename `Sample Test` to `Service Request`

- [ ] Move the `Request Sample test` to Actions on the right side. Its good to have page specfic actions also going into the right nav. (example `Assing to a Volunteer` should only be avilable in demography)

Mahendar0701 committed:
handled volunter tab, Rename Sample Test to Service Request, page specfic actions

Mahendar0701 commented:
@bodhish added requested changes. once look into it.

Can i do "displaying Shifiting details as a list" in my next PR , as new tab designs are proposed in https://www.figma.com/design/Z93EYKSa1MdBmXndsMBJSQ/Care?node-id=905-901&node-type=canvas&t=cA4CvHwv7gPc5Ij2-0  i will update all tab designs in my next PR. What do you say?

I want to close this PR becz its been so long...

Mahendar0701 committed:
test case

bodhish commented:
@Mahendar0701 consider this closed, @shivankacker will get this merged.

nihal467 committed:
Merge branch 'develop' into adjusting-margins-in-patient-details#8808

nihal467 commented:
@shivankacker @bodhish is it ready for QA testing, iam seeing a lot of recent commits and conversation,

shivankacker commented:
@nihal467 it is work in progress
